src/runtime.rs

```diff
@@ -634,16 +713,26 @@ pub(crate) struct StampedValue<V> {
     pub(crate) changed_at: Revision,
 }
 
-struct DependencyGraph<DB: Database> {
+#[derive(Debug)]
+struct Edge<K> {
+    id: RuntimeId,
+    path: Vec<K>,
```

Comment:
OK, so we are creating this vector eagerly (and repeatedly) for each cross-thread call. I'm *still* inclined to merge the PR, but I suspect this was one of the things that motivated me to want to examine an alternative impl strategy.

Reply:
For that matter, though, I still think the "parallel" support is kind of weak in salsa and an area for further development. =)

Reply:
True, been a while since I wrote it but since it takes the path from the local state (which is not thread safe) I couldn't think of a way to handle it lazily.